camel-core/src/main/java/org/apache/camel/processor/Throttler.java

```diff
@@ -111,13 +128,21 @@ public boolean process(final Exchange exchange, final AsyncCallback callback) {
                 throw new RejectedExecutionException("Run is not allowed");
             }
 
-            calculateAndSetMaxRequestsPerPeriod(exchange);
+            Integer key;
+            if (correlationExpression != null) {
+                key = correlationExpression.evaluate(exchange, Integer.class);
```

Comment:
If we use `key` as a finer grain lock, I would make a defensive copy because I'm not sure that the reference returned by evaluate could be used as a lock by another thread for another purpose.

Reply:
good point